### PR TITLE
CM-17814 - SCA commit range

### PR DESCRIPTION
**Pull request by avishaiamiel**

### CODE REVIEW

File: cli/code_scanner.py [hunk 1]
```diff
@@ -85,24 +85,26 @@ def scan_repository_commit_history(context: click.Context, path: str, commit_ran
 
 def scan_commit_range(context: click.Context, path: str, commit_range: str):
     scan_type = context.obj["scan_type"]
-
-    if scan_type != SECRET_SCAN_TYPE:
+    if scan_type != (SECRET_SCAN_TYPE and SCA_SCAN_TYPE):
```
Comment (MichalBor) [inline cli/code_scanner.py:88]: make a const list of COMMIT_RANGE_SCAN_SUPPORTED_SCAN_TYPES

File: cli/code_scanner.py [hunk 2]
```diff
@@ -85,24 +85,26 @@ def scan_repository_commit_history(context: click.Context, path: str, commit_ran
...
-            doc = Document(get_path_by_os(os.path.join(path, get_diff_file_path(blob))),
-                           blob.diff.decode('utf-8', errors='replace'), True, commit_id)
-            documents_to_scan.append(doc)
-
-            documents_to_scan = exclude_irrelevant_documents_to_scan(context, documents_to_scan)
-            logger.debug('Found all relevant files in commit %s',
-                         {'path': path, 'commit_range': commit_range, 'commit_id': commit_id})
-    return scan_documents(context, documents_to_scan, is_git_diff=True, is_commit_range=True)
+    if scan_type == SCA_SCAN_TYPE:
+        return scan_sca_commit_range(commit_range, context, path)
```
Comment (MichalBor) [inline cli/code_scanner.py:92]: move all the logic here to sca code scanner class

File: cli/code_scanner.py [hunk 3]
```diff
@@ -633,3 +723,8 @@ def _map_detections_per_file(detections) -> List[DetectionsPerFile]:
 
     return [DetectionsPerFile(file_name=file_name, detections=file_detections)
             for file_name, file_detections in detections_per_files.items()]
+
+
+def parse_commit_range(commit_range: str) -> (str, str):
+    commit_range_list = commit_range.split('...')
```
Comment (MichalBor) [inline cli/code_scanner.py:729]: commit range does not have to contain `...`

File: cli/code_scanner.py [hunk 4]
```diff
@@ -367,6 +438,26 @@ def exclude_detections_by_exclusions_configuration(scan_type: str, detections) -
     return [detection for detection in detections if not _should_exclude_detection(detection, exclusions)]
 
 
+def get_commit_range_modified_documents(path: str, from_commit_rev: str, to_commit_rev: str) -> (
+        List[Document], List[Document]):
+    from_commit_documents = []
+    to_commit_documents = []
+    repo = Repo(path)
+    diff = repo.commit(from_commit_rev).diff(to_commit_rev)
+    modified_files_diff = [change for change in diff if change.change_type != 'D']
```
Comment (MichalBor) [inline cli/code_scanner.py:447]: what is it "D"? move to const

File: cli/helpers/sca_utils.py [hunk 5]
```diff
@@ -0,0 +1,15 @@
+ecosystem_project_files = {
```
[the remark this answers is not in the document]
Comment (avishaiamiel) [reply cli/helpers/sca_utils.py:1]: what will happen when we will need to add helper for another scan type? you will create new folder for it also?
I dont see issue with putting all in the helpers for now but... your call :). let me know if you still want me to move it.

File: cli/code_scanner.py (hunk 3 above)
Comment (MaorDavidzon) [reply cli/code_scanner.py:729]: Let's move it to a separate command. Cause in the SCA is not really scanning a range of commits. It scans diff between 2 commits

File: cli/code_scanner.py [hunk 6]
```diff
@@ -367,6 +438,26 @@ def exclude_detections_by_exclusions_configuration(scan_type: str, detections) -
     return [detection for detection in detections if not _should_exclude_detection(detection, exclusions)]
 
 
+def get_commit_range_modified_documents(path: str, from_commit_rev: str, to_commit_rev: str) -> (
+        List[Document], List[Document]):
+    from_commit_documents = []
+    to_commit_documents = []
+    repo = Repo(path)
+    diff = repo.commit(from_commit_rev).diff(to_commit_rev)
+    modified_files_diff = [change for change in diff if change.change_type != COMMIT_DIFF_DELETED_FILE_IND]
```
Comment (MichalBor) [inline cli/code_scanner.py:447]: `COMMIT_DIFF_DELETED_FILE_CHANGE_TYPE`

File: cli/helpers/sca_utils.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (MichalBor) [reply cli/helpers/sca_utils.py:1]: its not sca_utils and its not helper, so sca_consts? or just add it to the consts file.
in addition change the name to `project_files_by_ecosystem_map` / `project_files_per_ecosystem_map`

File: cli/helpers/sca_code_scanner.py [hunk 7]
```diff
@@ -1,17 +1,59 @@
...
+    repo = Repo(path)
+    add_ecosystem_related_files_if_exists(from_commit_documents, repo, from_commit_rev)
+    add_ecosystem_related_files_if_exists(to_commit_documents, repo, to_commit_rev)
+
+
+def add_ecosystem_related_files_if_exists(documents: List[Document], repo: Repo, commit_rev: str):
+    documents_to_add: List[Document] = []
+    for doc in documents:
+        ecosystem = get_project_file_ecosystem(doc)
+        for ecosystem_project_file in sca_utils.ecosystem_project_files.get(ecosystem):
```
Comment (MichalBor) [inline cli/helpers/sca_code_scanner.py:28]: if ecosystem is null you will get exception here (because the value will be null)
please move

Comment (MichalBor) [inline cli/helpers/sca_code_scanner.py:28]: check if ecosystem not null
then move the code below to function
`get_ecosystem_project_files` -> return list of file to add and then do extend

File: cli/helpers/sca_code_scanner.py [hunk 8]
```diff
@@ -1,17 +1,59 @@
...
+                try:
+                    file_content = repo.git.show(f'{commit_rev}:{file_to_search}')
+                except GitCommandError:
+                    continue
+                documents_to_add.append(Document(file_to_search, file_content))
+
+    documents.extend(documents_to_add)
+
+
+def is_project_file_exists_in_documents(documents: List[Document], file: str):
```
Comment (MichalBor) [inline cli/helpers/sca_code_scanner.py:40]: use in `any` func

File: cli/helpers/sca_code_scanner.py (hunk 7 above)
Comment (avishaiamiel) [reply cli/helpers/sca_code_scanner.py:28]: move?

File: cli/helpers/sca_code_scanner.py [hunk 9]
```diff
@@ -1,17 +1,65 @@
...
+    add_ecosystem_related_files_if_exists(from_commit_documents, repo, from_commit_rev)
+    add_ecosystem_related_files_if_exists(to_commit_documents, repo, to_commit_rev)
+
+
+def add_ecosystem_related_files_if_exists(documents: List[Document], repo: Repo, commit_rev: str):
+    documents_to_add: List[Document] = []
+    for doc in documents:
+        ecosystem = get_project_file_ecosystem(doc)
+        if ecosystem is None:
+            logger.debug("failed to resolve project file ecosystem: %s", doc.path)
```
Comment (MichalBor) [inline cli/helpers/sca_code_scanner.py:29]: not failed - change to something like "document is not a ecosystem file"

File: cli/helpers/sca_code_scanner.py [hunk 10]
```diff
@@ -1,17 +1,65 @@
...
+            continue
+        documents_to_add = get_ecosystem_project_files_if_exists(documents, commit_rev, doc, ecosystem, repo)
```
Comment (MichalBor) [inline cli/helpers/sca_code_scanner.py:31]: add also debug log in case it ecosystem file

File: cli/helpers/sca_code_scanner.py [hunk 11]
```diff
@@ -1,17 +1,65 @@
...
+def get_ecosystem_project_files_if_exists(documents, commit_rev, doc, ecosystem, repo):
+    documents_to_add: List[Document] = []
+    for ecosystem_project_file in PROJECT_FILES_BY_ECOSYSTEM_MAP.get(ecosystem):
+        file_to_search = join_paths(get_file_dir(doc.path), ecosystem_project_file)
+        if not is_project_file_exists_in_documents(documents, file_to_search):
+            try:
+                file_content = repo.git.show(f'{commit_rev}:{file_to_search}')
+            except GitCommandError:
+                continue
+            documents_to_add.append(Document(file_to_search, file_content))
```
Comment (MichalBor) [inline cli/helpers/sca_code_scanner.py:45]: do you want to add also here a log?

File: cli/helpers/sca_code_scanner.py (hunk 9 above)
Comment (avishaiamiel) [reply cli/helpers/sca_code_scanner.py:29]: the expectation at this stage is that all documents are project files

File: cli/helpers/sca_code_scanner.py (hunk 10 above)
Comment (avishaiamiel) [reply cli/helpers/sca_code_scanner.py:31]: no need, we are printing all the files later on. print it here also will be redundant.

**Review by MichalBor — APPROVED** [review]
💪🏻

**Pull request merged**